library/std/src/sys/windows/fs.rs

```diff
@@ -696,13 +705,21 @@ impl<'a> Iterator for DirBuffIter<'a> {
         // last field (the file name) is unsized. So an offset has to be
         // used to get the file name slice.
         let (name, is_directory, next_entry) = unsafe {
+            let header_sz = mem::size_of::<c::FILE_ID_BOTH_DIR_INFO>();
+            assert!(buffer.len() >= header_sz);
             let info = buffer.as_ptr().cast::<c::FILE_ID_BOTH_DIR_INFO>();
-            let next_entry = (*info).NextEntryOffset as usize;
+            assert!(info.is_aligned());
+            let len_bytes = (*info).FileNameLength as usize;
+            // Note: we don't need `header_sz - 2` here, since the `[WCHAR; 1]`
+            // in the header and the trailing nul char (which always present but
+            // not counted in `len_bytes`) cancel each-other out.
+            assert!(buffer.len() >= header_sz + len_bytes);
```

Comment:
I think some of these asserts are a touch paranoid because we should be able to assume the OS is acting sensibly when filling the buffer. However, I don't think they do any harm either and it is possible they could catch programmer error or problems with miri's emulation.

Reply:
These checks are too annoying to write correctly without `offset_of!`, so I removed the ones not related to alignment.